### PR TITLE
Allow a device to be assigned to application

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## Description

Permit a device to be assigned to an application.

NOTE: Raised in draft to permit other engineers to test - before writing a whole bunch of tests and updating docs.

Notable:
1. when a device is added to an application, it is offered a fake snapshot with ID '0', computed (relevant) Env Vars and a modules object containing `"node-red": "latest"`.  This permits the device to start

#### Application level device implementation:
 - Developer Mode / Remote Access ✅
 - Device Logs ✅
 - Team Library ✅
 - Theme, device name and NR menu links ✅
 - Adding device in application view adds device to application ✅
 - Newly added device in application view updates status when config is applied to device and it starts up ✅
 - Upload Snapshot ❌
 - Download Snapshot ❌
 - Project Links ❌
 - Grouping and settings ❌
- Module cache for device at application level  ❌

#### TODO (this iteration):
 - [x] tests
 - [x] docs
 - [ ] manual testing combinations - to ensure there are no issues when running older platform with newer agent and vice versa
  
#### TODO (next iteration)
 - Provisioning Tokens for auto add to application
 - Module cache for device at application level
 - Support snapshots (upload/download)
 - Support grouping
 - Group settings (env/modules/etc)
 - Project Links (Device links? Concept?)

## Related Issue(s)

[Devices to be assigned to an Application, without being assigned to an Instance#2619](https://github.com/flowforge/flowforge/issues/2619)

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: docs/user/team/README.md
Comment (knolleary) [inline docs/user/team/README.md:68]:
```suggestion
| • Modify Environment Variables       | ✓     | ✓      | -      | -              |
| • Assign to/Remove from Application  | ✓     | -      | -      | -              |
| • Assign to/Remove from Instance     | ✓     | -      | -      | -              |
| • Delete Device                      | ✓     | -      | -      | -              |
```

Think application logically comes above Instance.

File: forge/comms/aclManager.js
Comment (knolleary) [inline forge/comms/aclManager.js:108]:
```suggestion
                // - ff/v1/+/a/+/command
```

File: forge/db/controllers/Device.js
Comment (knolleary) [inline forge/db/controllers/Device.js:93]: I wonder if we should set them blank rather than these non-blank values. I appreciate why you set it to `0` for the internal usage between platform and agent, but in the flows, I feel like a blank value would be clearer so we don't have to explain it away otherwise.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (knolleary) [inline forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:1]: This file needs to be renamed to ensure it is the most recent migration in the list.

At the time of commenting, the most recent migration is `20230818-01-add-access-token-name.js`

File: forge/routes/api/application.js
Comment (knolleary) [inline forge/routes/api/application.js:289]:
```suggestion
            tags: ['Applications'],
```

Not sure a separate section is needed - keeping it consistent with the /instances end point for eg

File: forge/routes/api/device.js
Comment (knolleary) [inline forge/routes/api/device.js:360]: Are these genuine todos or left overs?

File: forge/routes/api/device.js
Comment (knolleary) [inline forge/routes/api/device.js:699]: I assume this means 'todo later once we figure out how to handle application settings (if at all)' rather than needed for 1st iteration?

File: forge/routes/api/deviceLive.js
Comment (knolleary) [inline forge/routes/api/deviceLive.js:107]:
```suggestion
                        FF_SNAPSHOT_ID: '0',
                        FF_SNAPSHOT_NAME: 'None',
```

Same comment as elsewhere - I wonder if, when exposing in the flow, these should be blank rather than placeholders. Not sure.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (knolleary) [inline forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:19]: Does the migration need to also populate the `ApplicationId` column for all existing devices?

File: frontend/src/pages/device/dialogs/ModeChoiceDialog.vue
Comment (knolleary) [inline frontend/src/pages/device/dialogs/ModeChoiceDialog.vue:8]: I understand the intent, my concern is the user doesn't know how to make the device check in. Does the following help?

```suggestion
                        Please ensure the Device Agent is up to date and connected to the platform.
```

**Review by knolleary — CHANGES_REQUESTED** [review]
Looks good overall. Some comments and a couple bits that need updating - particularly around the migration.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (Steve-Mcl) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:19]: No. As per existing devices linked to an instance, the `project` object has its own relationship to Application - this remains.

The presence of this property is to signify a device is lined to an application not a project.

File: forge/db/controllers/Device.js
Comment (Steve-Mcl) [reply forge/db/controllers/Device.js:93]: This is _very_ temporary Nick. 

Happy to use a name or something other than 'None'

Are you proposing using `Blank` or `Something-else` as the name?

Are you also proposing using something other than `0` for the ID?  I was torn between somthing like `DEFAULT` ot `APPLICATION-SNAPSHOT` but ultimately, the temporary nature of this, I went with '0'

File: forge/routes/api/application.js
Comment (Steve-Mcl) [reply forge/routes/api/application.js:289]: Should this then not be `['Devices']` as it returns devices?

Not sure now that I understood the relevance of tags 🤔

File: forge/routes/api/device.js
Comment (Steve-Mcl) [reply forge/routes/api/device.js:360]: Future TODOs - in line with other TODOs already in surrounding tests.

TBH, they can be removed as, strictly speaking,  they are really separate unit test related to audit logging.

File: frontend/src/pages/device/dialogs/ModeChoiceDialog.vue
Comment (Steve-Mcl) [reply frontend/src/pages/device/dialogs/ModeChoiceDialog.vue:8]: better :)

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (Steve-Mcl) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:1]: @knolleary resolved in [9f9b53f](https://github.com/flowforge/flowforge/pull/2621/commits/9f9b53f46ef545d82102e3ff9974ec6725ca456f)

File: forge/db/controllers/Device.js
Comment (knolleary) [reply forge/db/controllers/Device.js:93]: In terms of the env vars on the device, I was meaning literally blank - ''.

File: forge/routes/api/application.js
Comment (knolleary) [reply forge/routes/api/application.js:289]: The tags determine where it shows up in the swagger doc output - https://app.flowforge.com/api/static/index.html

This is an application based API so should be under applications.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (knolleary) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:19]: Does this mean the `/applications/<id>/devices` endpoint only returns the devices assigned to just the application? I was working on the assumption it returned the superset of devices assigned to the application directly and devices assigned to instances within the application?

File: forge/db/controllers/Device.js
Comment (Steve-Mcl) [reply forge/db/controllers/Device.js:93]: hmmm. Would have to double check the logic across platform and device. 

For example, `'0' is used as a catch point in the device logic to know that A) the snapshot is _something_ and B) it is a _temporary/starter_snapshot for the device.  Existing logic for checking there IS a snapshot then works as is (because `.id` of '0' is _something_ and it signifies an application starter snapshot)

I do suspect this would cause some inconsistencies as it stands right now.

Happy to change - or move forward as is.  Please advise.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (Steve-Mcl) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:19]: Apologies if that has been lost in translation. The current implementation is literal i.e. "Allow a Device to be assigned to Application" as per my understanding and interpretation. 

As it happens, I am not totally sold on your counter proposal - if feels odd somehow?  Perhaps with a rejig of the UI that might feel right but in the current form, it more _feels_ right to me. 

And of course, as this all changes again when grouping is introduced, I/you might feel different about one vs other? And anyhow, the design stage should route this out and I will be sure to ask for clarification to minimise confusion.

File: forge/db/controllers/Device.js
Comment (knolleary) [reply forge/db/controllers/Device.js:93]: I would have chosen to leave them blank as there is no running snapshot and I don't think the scenario of an existing device using logic based on this env var wanting to migrate to this new mode instantly (without snapshots in place) is really enough of a concern.

But I'm not going to push this.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (knolleary) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:19]: It's a matter of perspective. Instances are owned by an Application. If a Device is assigned to an instance then it is 'owned' by the parent Application. So a view of devices in the application would include those that have been further assigned to individual instances. But regardless, as I said, we can proceed with what you have and I'll write up a proposed set of changes next week we can discuss in the broader context.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (knolleary) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:1]: Not resolved - we now how two migration files :) Need to delete the old one.

**Review by knolleary — CHANGES_REQUESTED** [review]

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (Steve-Mcl) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:1]: oh, odd, i did a rename of file and (wrongly) expected that to remove the old one. 

Will fix now.

**Comment by knolleary** [comment]
I may have missed it, but is there any guarding against assigning an older device agent to an application? This is only going to work with agents that are running with https://github.com/flowforge/flowforge-device-agent/pull/161

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (Steve-Mcl) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:1]: Double odd

image

Comment (Steve-Mcl) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js:1]: Doh! 👐 

image

**Comment by Steve-Mcl** [comment]
> I may have missed it, but is there any guarding against assigning an older device agent to an application

Yeah, there is an agent version check.  And the reason for updating the vue page comment about "Device needs to check in" - so that it knows it is permitted to assign the device to an application.

let me double verify - 5 mins.

**Comment by Steve-Mcl** [comment]
> > I may have missed it, but is there any guarding against assigning an older device agent to an application
> 
> Yeah, there is an agent version check. And the reason for updating the vue page comment about "Device needs to check in" - so that it knows it is permitted to assign the device to an application.
> 
> let me double verify - 5 mins.

Urg, seems I lost that bit of work. 

please hold off until I find / merge / re-add.

**Comment by knolleary** [comment]
Hmm, I can no longer find any trace of the conversation thread in this PR related to the `applications/id/devices` api... which is going to make it hard to refer back to.

Adding a note that, whilst reviewing Device Agent changes, I can see lots of assumptions based on `application` only being set if assigned to an application and not an instance. So any future change that sees `application` set if assigned to either the app or an instance within the application will be hard to retrofit.

File: forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js
Comment (knolleary) [reply forge/db/migrations/20230816-01-add-ApplicationId-to-Devices.js]: Looks good - resolving this thread.

**Comment by Steve-Mcl** [comment]
@knolleary pretty sure this is now ready for re-review. Had some issue with merge conflict after fixing tests locally & pushing  but this should all be resolved now.

**Comment by Steve-Mcl** [comment]
small buglet found while testing the changes requested in the partner PR on the device agent.

Resolving this now & pushing update to agent imminently.

File: frontend/src/pages/team/Devices/index.vue
Comment (knolleary) [inline frontend/src/pages/team/Devices/index.vue:15]:
```suggestion
                    <p>FlowFuse can be used to manage instances of Node-RED running on remote devices.</p>
                    <p>Each device must run the <a href="https://flowfuse.com/docs/user/devices/" target="_blank">FlowFuse Device Agent</a>, which connects back to the platform to receive updates.</p>
                    <p>Devices are registered to a Team, and assigned to an Application or an Instance.</p>
                    <p>Flows can then be deployed remotely to the devices as a Snapshot.</p>
```

**Review by knolleary — APPROVED** [review]

**Pull request merged**
